### ISSUE TITLE
Feature proposal: "Add to selection" support

### ISSUE THREAD

**Issue by petrcermak** [catapult-project/catapult#1529]
I'm currently improving the analysis UI for multiple memory dumps (#1369). One of the main use cases will be to analyze how memory usage changed between two timestamps (i.e. diff). Since all manual selections in Trace Viewer must be contiguous (i.e. without "gaps"), the only way to provide a diff between two non-adjacent memory dumps is to make the user select the whole range of dumps between the two timestamps, which is not user-friendly.

Therefore, I propose we **add support for adding events to the current selection**. The most natural thing would probably be (this seems to be the de facto standard):

   * Click/drag a rectangle: Set new selection (existing behavior)
   * _Ctrl_ + click/drag a rectangle: Add new selection to the current selection (new behavior)

I think this would be very simple to implement. The only issue I see here is that we already use _Ctrl_ to toggle mouse mode to "zoom". I don't know how many people actually know/use it (I suppose only a couple of power users?). My suggestion would be to change the key for the "zoom" mode to _Alt_.

@natduca, @dj2: WDYT?

**Comment by natduca** [comment]
Love it.

**Comment by dj2** [comment]
sgtm.

**Comment by petrcermak** [comment]
I've just uploaded a prototype of the feature: https://codereview.chromium.org/1396473004/. Any feedback is welcome.

I added a <tt>setModifierFlag</tt> method to <tt>\<tr-ui-b-mouse-mode-selector\></tt>. This allows the timeline view to specify that it wants the <tt>beginselection</tt> event to have an <tt>appendselection</tt> flag when Ctrl/Cmd is pressed:

```
mouseModeSelector.setModifierFlag(
    'beginselection' /* event */,
    tr.ui.b.MODIFIER.CMD_OR_CTRL,
    'appendselection' /* flag */);
```

Alternatively, this could be pushed into the mouse mode selector and the user (timeline track view here) could just specify which flags it allows. I think that might actually be better because it would allow us to temporarily change the "selection" button (pointer) to an "add selection" button (pointer with a plus sign) when you press Ctrl.

The only issue here is that we would now be using modifiers (Ctrl/Cmd, Shift, Space) for two different purposes: (1) to *completely change* mode and (2) to *slightly modify* the current mode. I find this a little unfortunate as it's difficult to reason about (for us) and confusing to use (for our users). I can see several ways out of this:

  1. **Don't add the new feature** (status quo). I think this would be a shame because there would be way to add events to the current selection.
  2. **Merge the two frameworks**. Adding an event to the current selection could in theory be just a separate mouse mode (i.e. another button on the floating mouse mode selector). The issue I see with this is that it will make the mouse selector even bigger. If we want to add any other modifiers (e.g. for selection intersection and/or selection subtraction), it would become huge (and it already somehow manages to cover the part of the trace that you are interested in).
  3. **Drop support for changing mouse mode via modifiers**. I have personally never used Ctrl/Cmd/Shift/Space to temporarily change the mouse mode. So my question is: **Does anyone actually use this?** If not, then I would be strongly in favor of this option.
